components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultAuthenticationRequestHandler.java

```diff
@@ -443,6 +455,62 @@ protected void concludeFlow(HttpServletRequest request, HttpServletResponse resp
         sendResponse(request, response, context);
     }
 
+    /**
+     * Method used to store user and session related data to the database.
+     *
+     * @param context           {@link AuthenticationContext} object with the authentication request related data
+     * @param sessionContextKey of the authenticated session
+     */
+    private void storeSessionData(AuthenticationContext context, String sessionContextKey)
+            throws UserSessionException {
+
+        for (AuthenticatedIdPData authenticatedIdPData : context.getCurrentAuthenticatedIdPs().values()) {
+            String userName = authenticatedIdPData.getUser().getUserName();
+            String tenantDomain = getAuthenticatedUserTenantDomain(context, null);
+            int tenantId = (tenantDomain == null) ? MultitenantConstants.INVALID_TENANT_ID : IdentityTenantUtil
+                    .getTenantId(tenantDomain);
+            String userStoreDomain = authenticatedIdPData.getUser().getUserStoreDomain();
+            String idpName = authenticatedIdPData.getIdpName();
+            String userId;
+            try {
+                int idpId = UserSessionStore.getInstance().getIdPId(idpName);
+                userId = UserSessionStore.getInstance().getUserId(userName, tenantId, userStoreDomain, idpId);
+
+                // This synchronized block is added to handle a situation when new user try to get authenticated
+                // from multiple nodes at the same time (which is rare case).
+                if (userId == null) {
```

Comment:
Is this if clause necessary for synchronization?
inside the synchronize block it is anyway checking the userId== null